crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+
+use core::panic;
+use std::{sync::Arc, time::Duration};
+
+use broadcast::Receiver;
+use futures::stream::FuturesOrdered;
+use sui_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest},
+    committee::StakeUnit,
+    error::{SuiError, SuiResult},
+    messages::TransactionEffects,
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+const TASKS_PER_CORE: usize = 1;
+
+pub struct CheckpointExecutor {
+    mailbox: Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_executed_checkpoint: Option<CheckpointSequenceNumber>,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        mailbox: Receiver<VerifiedCheckpoint>,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        let highest_executed_checkpoint =
+            checkpoint_store.get_highest_executed_checkpoint_seq_number()?;
+
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            highest_executed_checkpoint,
+            authority_state,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        let mut finished: FuturesOrdered<JoinHandle<(u64, VerifiedCheckpoint)>> =
+            FuturesOrdered::new();
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
+
+        // In memory only "watermark" to keep track of the highest checkpoint for which we've
+        // spun up an execution task. This is to prevent us from rescheduling the same checkpoints
+        // in the case where we receive a synced checkpoint message before scheduled checkpoints have
+        // completed, and hence have had the chance to update the on-disk watermark `highest_executed_checkpoint`.
+        let mut highest_scheduled_checkpoint: Option<u64> = None;
+
+        loop {
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((seq_num, checkpoint))) = finished.next() => {
+                    self.highest_executed_checkpoint = Some(seq_num as CheckpointSequenceNumber);
```

Comment:
Given we already have this watermark in the persistent store i'm not sure what the benefit of having the same watermark in memory would be, other than maybe not having to always read it from the store.